### PR TITLE
Super Mario Sunshine: Specular lighting, new objects, scenario names

### PR DESCRIPTION
**Pull request by impiaaa**

### CODE REVIEW

**Review by magcius — APPROVED** [review]

File: src/Common/JSYSTEM/J3D/J3DLoader.ts [hunk 1]
```diff
@@ -1037,12 +1037,10 @@ function readMAT3Chunk(buffer: ArrayBufferSlice): MAT3 {
...
             const attenuationFunction: GX.AttenuationFunction = (
-                attnFn === 0 ? GX.AttenuationFunction.NONE :
-                attnFn === 1 ? GX.AttenuationFunction.SPEC :
-                attnFn === 2 ? GX.AttenuationFunction.NONE :
-                attnFn === 3 ? GX.AttenuationFunction.SPOT : -1
+                attnFn === 0 ? GX.AttenuationFunction.SPEC :
+                attnFn === 1 ? GX.AttenuationFunction.SPOT :
+                               GX.AttenuationFunction.NONE
             );
-            assert((attenuationFunction as number) !== -1);
```
[the remark this answers is not in the document]
Comment (impiaaa) [reply src/Common/JSYSTEM/J3D/J3DLoader.ts:1045]: In `JSystem::calcColorChanID`, the two attenuation function bits are calculated as:

    ((attenuationFunction != 2) * 0x200) | ((attenuationFunction != 0) * 0x400)

which results in Spec for 0, Dir/None for 2, and Spot for anything else.[^1]

libogc's [`GX_SetChanCtrl`](/devkitPro/libogc/blob/9c16b6af30a105c51433bd76efed7d9197baa868/libogc/gx.c#L2189-L2205) calculates those bits as:

    ((attn_fn < 2) * 0x200) | ((attn_fn > 0) * 0x400)

which results in Spec for anything less than 1, Spot for 1, and Dir/None for anything greater than 1.

I will admit that I had not actually fully analyzed `calcColorChanID` when I made this change, instead using a function that matched the results of `GX_SetChanCtrl` for 0, 1, 2, and 3. Should I make another change to bring `readColorChannel` closer to one of the above two functions?

[^1]: The previous implementation for `readColorChannel` used a lookup table found in `J3DColorBlockLightOff::load`. However, that table is for the *reverse* operation, to go from register value to GX enum (Spec to 0, Spot to 1, and None and Dir to 2). That enum value is then mapped back to a register value *again* in `JSystem::J3DGDSetChanCtrl`, in the same way as `JSystem::calcColorChanID`.

File: src/j3d/sms_scenes.ts [hunk 2]
```diff
@@ -568,142 +755,388 @@ export class SunshineSceneDesc implements Viewer.SceneDesc {
...
-        } else if (modelEntry.s !== undefined) {
-            scene = modelEntry.s();
+            const jpaFilename = modelEntry.p.toLowerCase();
+            const jpaFile = assertExists(rarc.findFile(jpaFilename), jpaFilename);
+            const jpaData = lookupEffect(jpaFile);
         }
 
-        if (scene === null)
+        if (scene === null) {
+            console.log("couldn't load "+JSON.stringify(modelEntry));
```
[the remark this answers is not in the document]
Comment (impiaaa) [reply src/j3d/sms_scenes.ts:1137]: Good to know, though `modelEntry` here is just a little model information, something like `{ k: 'water_roll_block', m: 'water_roll_block.bmd' }`—nothing really worth poking at :slightly_smiling_face:

File: src/Common/JSYSTEM/J3D/J3DLoader.ts [hunk 3]
```diff
@@ -886,21 +886,25 @@ function readMAT3Chunk(buffer: ArrayBufferSlice): MAT3 {
...
-            const rasSwapC = view.getUint8(tevSwapModeTableInfoOffset + tevSwapModeTableRasIndex * 0x04 + 0x02);
-            const rasSwapD = view.getUint8(tevSwapModeTableInfoOffset + tevSwapModeTableRasIndex * 0x04 + 0x03);
-            const texSwapA = view.getUint8(tevSwapModeTableInfoOffset + tevSwapModeTableTexIndex * 0x04 + 0x00);
-            const texSwapB = view.getUint8(tevSwapModeTableInfoOffset + tevSwapModeTableTexIndex * 0x04 + 0x01);
-            const texSwapC = view.getUint8(tevSwapModeTableInfoOffset + tevSwapModeTableTexIndex * 0x04 + 0x02);
-            const texSwapD = view.getUint8(tevSwapModeTableInfoOffset + tevSwapModeTableTexIndex * 0x04 + 0x03);
-
-            const rasSwapTable = [rasSwapA, rasSwapB, rasSwapC, rasSwapD] as const;
-            const texSwapTable = [texSwapA, texSwapB, texSwapC, texSwapD] as const;
+            let rasSwapTable: readonly [number, number, number, number] = [0, 1, 2, 3] as const;
```
[the remark this answers is not in the document]
Comment (impiaaa) [reply src/Common/JSYSTEM/J3D/J3DLoader.ts:889]: Ah, sorry, I didn't know about that.

File: src/j3d/sms_scenes.ts [hunk 4]
```diff
@@ -273,17 +274,42 @@ export const enum SMSPass {
...
-function setGXLight(dst: Light, src: SceneBinObjLight, camera: Camera): void {
-    lightSetWorldPosition(dst, camera, src.x, src.y, src.z);
-    colorFromRGBA(dst.Color, src.r/0xFF, src.g/0xFF, src.b/0xFF, src.a/0xFF);
-}
-
 class LightConfig {
     public lightObj: SceneBinObjLight[] = [];
 
+    private LARGE_NUMBER = -1048576.0;
+    private initSpecularDir(lit: Light, nx: number, ny: number, nz: number) {
```
[the remark this answers is not in the document]
Comment (impiaaa) [reply src/j3d/sms_scenes.ts:464]: Sure, would you like me to bring it out? possibly into a `lightSetWorldSomething` in gx_material.ts (or maybe an equivalent `lightSetWorldSomething` is already in gx_material.ts and I haven't noticed)

File: src/Common/JSYSTEM/J3D/J3DLoader.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (magcius) [reply src/Common/JSYSTEM/J3D/J3DLoader.ts:1045]: Thank you, I'm happy with this explanation now!

File: src/j3d/sms_scenes.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (magcius) [reply src/j3d/sms_scenes.ts:464]: There's already `lightSetWorldDirection`, however, looking at the code, I might change how it works.

[the remark this answers is not in the document]
Comment (magcius) [reply src/j3d/sms_scenes.ts:464]: There's already `lightSetWorldDirection`, however, looking at the code, I might change how it works. I just meant pull this out to be a function shared between LightConfig and the other copy/paste of this function in the file.

**Pull request merged**